### DISCUSSION TITLE
Order filter form: payment provider filter includes non-completed payments

### DISCUSSION THREAD

**Discussion by stapelberg** [pretix/pretix#4508] [Ideas]
### Problem and impact

I’m querying a URL like the following so that I can list all orders of 2024Q1 that were paid via a specific payment method:

http://localhost:8000/control/event/foo/bar/orders/?expert-status=pv&expert-subevents_from_0=01.01.2024&expert-subevents_from_1=00%3A00%3A00&expert-subevents_to_0=01.04.2024&expert-subevents_to_1=00%3A00%3A00&expert-created_from_0=&expert-created_from_1=&expert-created_to_0=&expert-created_to_1=&expert-email=&expert-comment=&expert-locale=&expert-email_known_to_work=unknown&expert-total=&expert-payment_sum_min=&expert-payment_sum_max=&expert-sales_channel=&expert-has_checkin=unknown&expert-item=5&expert-provider=giftcard&expert-invoice_address_company=&expert-invoice_address_name=&expert-invoice_address_street=&expert-invoice_address_zipcode=&expert-invoice_address_city=&expert-invoice_address_country=&expert-attendee_name=&expert-attendee_email=&expert-attendee_address_company=&expert-attendee_address_street=&expert-attendee_address_zipcode=&expert-attendee_address_city=&expert-attendee_address_country=&expert-ticket_secret=

In words: Suchanfrage: Bestellstatus: None · Alle Termine beginnend am oder nach: 01.01.2024 00:00 · Alle Termine beginnend vor: 01.04.2024 00:00 · Produkte: Merch-Bestellung · Zahlungsmethode: Wertgutschein / Geschenkgutschein

I noticed that I have a certain order which shows up when filtering for `expert-provider=giftcard` **and** when filtering for `expert-provider=datatrans`.

A closer look revealed that the order in question was *attempted* with datatrans, but then switched to giftcard, which was the payment method that was ultimately used:
![2024-08-11-pretix-bug](https://github.com/user-attachments/assets/60da9c15-935d-49da-8f7b-0aeea48313d6)

### Expected behaviour

The payment provider filter should not consider canceled payments, only successful payments.

### Browser (software, desktop or mobile?) and version

Chrome

### Operating system, dependency versions

Linux

### Version

2024.7

**Comment by raphaelm** [comment]
I think this is expected behavior and also useful in some situations: Someone claims payment method X is not working, how are you to find their attempt and the error message in a large data set?

Side note: Filtering orders by payment method is useless for many purposes people try to use it for, since orders can be paid with mixed payment methods.

**Comment by stapelberg** [comment]
> I think this is expected behavior and also useful in some situations: Someone claims payment method X is not working, how are you to find their attempt and the error message in a large data set?

By entering their name and looking at their orders :)

If this is the intended use-case, a separate view of payments (or payment method aggregate stats) would be better, I think.

But including attempted payments when filtering still seems surprising and not useful to me.

> Side note: Filtering orders by payment method is useless for many purposes people try to use it for, since orders can be paid with mixed payment methods.

Yes, I’m aware. Luckily such orders are rare for us. The goal of fixing this issue is to reduce the number of “special” orders shown in the pretix overview, which will make verifying our accounting easier.

BTW, can one disable accepting multiple payment methods for a single order?

**Comment by raphaelm** [comment]
> If this is the intended use-case, a separate view of payments (or payment method aggregate stats) would be better, I think.

We actually added that on global level: /control/search/payments/

But just as well I could ask why you don't use the "Payments & refunds" export as an exact way to solve your problem ;) 

> But including attempted payments when filtering still seems surprising and not useful to me.

I understand, but I also don't think your patch fixes this without causing new surprises.

Your search query is "status = paid + payment = datatrans". Your PR fixes that query.

Another very reasonable query is "status = pending + payment = banktransfer" to check which payments to look out for on the bank account. This query would be less useful than now with your change.

I understand your pain, but I fail to see the obvious way forward.

> BTW, can one disable accepting multiple payment methods for a single order?

No, that would break sooooo many things
- redeeming gift cards that do not have enough value for the entire order
- switching payment methods after a ticket upgrade for the remainder
- incoming bank transfers with insufficient value, then a credit card payment for the remainder
- partial refunds (if we extend the disabling to refunds)

**Comment by stapelberg** [comment]
> We actually added that on global level: /control/search/payments/

Ah, great!
 
> But just as well I could ask why you don't use the "Payments & refunds" export as an exact way to solve your problem ;)

Yes, the data from the “Payments & refunds” export seems better suited on first glance, but the presentation (excel/csv export that I need to post-process) is not well-suited for my purpose.

Also, the export lacks the filters I need. We have two interesting timestamps for each order, the payment date and the event date. We have multiple ledgers to compare against: When checking for correctness, we use the event date (we have a different log from a different system which contains entries for each event). When checking our accounting, we use the payment date. 

Currently, I have some custom self-written code which does the accounting, and I’m linking to prepared pretix queries as a way to sanity check what my code prints. For example, if my code says it considered 347 pretix orders, but clicking the link yields a list of 348 orders, I know there is a bug somewhere.

AFAICT, only the order filter view allows me to filter by (all of):
* product
* payment date
* event date
* payment mechanism

So, for example, my code will prepare a quarterly earnings statement and link to an order filter view for orders that were paid this quarter, but whose event is next quarter, and which came in via payment method datatrans.

> Your search query is "status = paid + payment = datatrans". Your PR fixes that query.
> 
> Another very reasonable query is "status = pending + payment = banktransfer" to check which payments to look out for on the bank account. This query would be less useful than now with your change.
> 
> I understand your pain, but I fail to see the obvious way forward.

Yeah, that’s a good example. Maybe the payment filter could be split into two to satisfy all the use-cases? payment_any and payment_completed? 

> > BTW, can one disable accepting multiple payment methods for a single order?
> 
> No, that would break sooooo many things
> 
> * redeeming gift cards that do not have enough value for the entire order
> * switching payment methods after a ticket upgrade for the remainder
> * incoming bank transfers with insufficient value, then a credit card payment for the remainder
> * partial refunds (if we extend the disabling to refunds)

Yes, I realize that all these actions would no longer work, but that’s exactly what I want :) 

Each of these cases makes our accounting more complicated and is very rare in our operation (less than once a quarter). We actually would prefer if customers had to cancel and re-order, sticking to a single payment method.

No big deal if this would be terribly complicated to implement, just wanted to point out that flexibility is great for many use-cases, but might not be desired in all use-cases :)
